src/Agent.Worker/TestResults/TestRunPublisher.cs

```diff
@@ -154,6 +160,34 @@ public TestRunData ReadResultsFromFile(TestRunContext runContext, string filePat
         }
         #endregion
 
+        private bool HierarchicalSubresultPreProcessing(string automatedTestName, List<TestCaseSubResultData> subResults, int level = 1)
+        {
+            int maxSubResultHierarchyLevel = TestManagementConstants.maxHierarchyLevelForSubresults;
+            int maxSubResultIterationCount = TestManagementConstants.maxSubResultPerLevel;
+            if (subResults == null || !subResults.Any())
+            {
+                return true;
+            }
+            if (level > maxSubResultHierarchyLevel)
+            {
+                return false;
+            }
+            if (subResults.Count > maxSubResultIterationCount)
+            {
+                _executionContext.Warning(StringUtil.Loc("MaxSubResultLimitReached", automatedTestName, maxSubResultIterationCount));
+                subResults.RemoveRange(maxSubResultIterationCount, subResults.Count - maxSubResultIterationCount);
+            }
+            foreach (var subresult in subResults)
+            {
+                if (HierarchicalSubresultPreProcessing(automatedTestName, subresult.SubResultData, level + 1) == false)
+                {
+                    _executionContext.Warning(StringUtil.Loc("MaxHierarchyLevelReached", maxSubResultHierarchyLevel));
```

Comment:
since you are calling the methods recursive, this warning might print out multiple times, is that what you want?